In [1]:
import pandas as pd
import numpy as np

In [2]:
df=pd.read_csv(r"C:\Users\PC\Documents\Data science\insurance_project.csv")  #Importing data set

In [3]:
df

Unnamed: 0       age  sex       bmi  children  smoker  region   charges
0              0  0.021739    0  0.387788       0.0       1       3  0.472641
1              1  0.000000    1  0.578435       1.0       0       2  0.018101
2              2  0.217391    1  0.553426       3.0       0       2  0.099775
3              3  0.326087    1  0.219065       0.0       0       1  0.625547
4              4  0.304348    1  0.419617       0.0       0       1  0.082306
...          ...       ...  ...       ...       ...     ...     ...       ...
1188        1334  0.000000    0  0.518350       0.0       0       0  0.032506
1189        1335  0.000000    0  0.678467       0.0       0       2  0.015231
1190        1336  0.065217    0  0.319584       0.0       0       3  0.026568
1191        1337  0.934783    0  0.425788       0.0       1       1  0.840140
1192        1338  0.461043    1  0.477538       0.0       1       2  0.364264

[1193 rows x 8 columns]

In [4]:
# Part 1


#Analysing the data set, i realized that it is a regression problem.
#the dependent variable here is the charges, which is a continuous quantity



In [5]:
y=df['charges']    #Dependent variable

In [6]:
y

0       0.472641
1       0.018101
2       0.099775
3       0.625547
4       0.082306
          ...   
1188    0.032506
1189    0.015231
1190    0.026568
1191    0.840140
1192    0.364264
Name: charges, Length: 1193, dtype: float64

In [7]:
x=df[['age','children','smoker','region']] # Independent variables
                                            #NOTE: I have used the 4 best features from my insurance data set for model building
                                            # i identified these features during EDA

In [8]:
x

age  children  smoker  region
0     0.021739       0.0       1       3
1     0.000000       1.0       0       2
2     0.217391       3.0       0       2
3     0.326087       0.0       0       1
4     0.304348       0.0       0       1
...        ...       ...     ...     ...
1188  0.000000       0.0       0       0
1189  0.000000       0.0       0       2
1190  0.065217       0.0       0       3
1191  0.934783       0.0       1       1
1192  0.461043       0.0       1       2

[1193 rows x 4 columns]

In [9]:
#splitting data into training and testing data sets

from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y,test_size=0.2, random_state=4)

# Building a Linear Regression Model

In [91]:
# Creating an instance of LinearRegression

from sklearn.linear_model import LinearRegression
R=LinearRegression()

In [11]:
R.fit(x_train,y_train)  # Building the model using training data set

LinearRegression()

In [12]:
R.intercept_

0.07317763083972267

In [13]:
R.coef_     #coefficients of my independent variables (0.33065752,  0.01309251,  0.43380983, -0.01546341)

array([ 0.33065752,  0.01309251,  0.43380983, -0.01546341])

In [14]:
# model equation: charges = age*0.33065752 + children*0.01309251 +smoker*0.43380983 - region*0.01546341 + 0.07317763083972267

In [15]:
y_predict=R.predict(x_test) #Testing the model using the unseen data

# Evaluating the model using different metrics

In [16]:
from sklearn import metrics as m

In [17]:
m.mean_squared_error(y_test,y_predict)  #evaluating our model using the mean squared error

0.01786517011817138

In [18]:
np.sqrt(m.mean_squared_error(y_test,y_predict))  #evaluating our model using the Root mean squared error (RMSE)

0.1336606528420813

In [19]:
m.r2_score(y_test,y_predict)   # model accuracy

0.6436300523736447

In [20]:
linear_score=m.r2_score(y_test,y_predict)

In [21]:
y_train_predict=R.predict(x_train) #checking how our model is fairing using seen data

In [22]:
m.mean_squared_error(y_train,y_train_predict)

0.01926460123638254

In [23]:
np.sqrt(m.mean_squared_error(y_train,y_train_predict)) #Bias of our model using the seen data

0.13879697848434072

In [24]:
m.r2_score(y_train,y_train_predict) #accuracy of model for our training data set

0.5844764237552935

In [25]:
print('Variance: ',0.13879697848434072-0.1336606528420813) #variance between the training and testing data set
                                                

Variance:  0.0051363256422594095


# Building The Ridge Regression Model

In [26]:
# Ridge Regression

from sklearn.linear_model import Ridge
from sklearn.linear_model import Lasso
from sklearn.metrics import r2_score

In [27]:
Ridge_R = Ridge(alpha=0.01)

In [28]:
Ridge_R.fit(x_train, y_train)  #Ridge Regression model building algorithm

Ridge(alpha=0.01)

In [29]:
y_train_pred_rr = Ridge_R.predict(x_train) #training the data with the seen data

In [30]:
m.mean_squared_error(y_train,y_train_pred_rr)  #Evaluating the models using mse

0.019264601607124805

In [31]:
np.sqrt(m.mean_squared_error(y_train,y_train_pred_rr)) #Bias of our model using the training data set

0.13879697981989667

In [32]:
r2_score(y_train,y_train_pred_rr) 

0.5844764157586497

In [33]:
y_test_pred_rr=Ridge_R.predict(x_test)  #using our test data set to see how our model is acting

In [34]:
m.mean_squared_error(y_test,y_test_pred_rr)

0.01786512421138697

In [35]:
np.sqrt(m.mean_squared_error(y_test,y_test_pred_rr)) 


0.13366048111310602

In [36]:
r2_score(y_test,y_test_pred_rr)  #accuracy of our model using the testing data 

0.6436309681107044

In [37]:
ridge_score=r2_score(y_test,y_test_pred_rr)

In [38]:
Ridge_R.intercept_

0.07320254574145427

In [39]:
Ridge_R.coef_

array([ 0.33061638,  0.01309255,  0.43375963, -0.01546399])

In [40]:
#Ridge Regression model: cherges=age*0.33061638 + children*0.01309255 + smoker*0.43375963 *region*0.01546399 + 0.07320254574145427

In [41]:
print("variance: ",0.13879697981989667 - 0.13366048111310602) # Variance for the ridge regression model

variance:  0.005136498706790643


# Building the Lasso Regression Model

In [42]:
#Lasso Regression

Lasso_R=Lasso(0.01)  #creating an instance of the Lasso Regressor

In [43]:
Lasso_R.fit(x_train,y_train) #model building algorithm using training data

Lasso(alpha=0.01)

In [44]:
y_train_pred_ls=Lasso_R.predict(x_train) #checking model using seen data

In [45]:
np.sqrt(m.mean_squared_error(y_train, y_train_pred_ls))

0.14716036817213274

In [46]:
r2_score(y_train,y_train_pred_ls) #model accuracy

0.5328919222644574

In [47]:
y_test_pred_ls = Lasso_R.predict(x_test)  #checking how the model is acting using the test data

In [48]:
np.sqrt(m.mean_squared_error(y_test,y_test_pred_ls))

0.14332340277445854

In [49]:
r2_score(y_test,y_test_pred_ls) #accuracy of  model

0.5902413423653934

In [50]:
lasso_score=r2_score(y_test,y_test_pred_ls)

In [51]:
Lasso_R.intercept_

0.13380205506502524

In [52]:
Lasso_R.coef_

array([ 0.21672156,  0.00666662,  0.32153698, -0.00894106])

In [53]:
#model of the lasso regression : charges= age*0.21672156 + children*0.00666662 + smoker*0.32153698-region*0.00894106 +0.1338020550650252

In [54]:
print("Variance: ",0.14716036817213274 - 0.14332340277445854  )

Variance:  0.003836965397674197



# Building The Decision Tree Model

In [55]:
from sklearn.tree import DecisionTreeRegressor

In [56]:
D_T_R = DecisionTreeRegressor(random_state=42)  #creating an instance of the decision tree

In [57]:
D_T_R.fit(x_train,y_train)  #Building the model using training data set

DecisionTreeRegressor(random_state=42)

In [58]:
y_train_pred_dtr=D_T_R.predict(x_train) # using the training data to see how our model is fairing

In [59]:
np.sqrt(m.mean_squared_error(y_train, y_train_pred_dtr)) #evaluating our model using training data set

0.08276835709905783

In [60]:
r2_score(y_train,y_train_pred_dtr)

0.8522374709038588

In [61]:
y_test_pred_dtr=D_T_R.predict(x_test) # testing our model using the testing data set

In [62]:
np.sqrt(m.mean_squared_error(y_test, y_test_pred_dtr))  # Bias of the testing data

0.1832329076652522

In [63]:
r2_score(y_test, y_test_pred_dtr)  #accuracy of our model using the unseen data

0.33026828731009505

In [64]:
DecisionT_score=r2_score(y_test, y_test_pred_dtr)

In [65]:
print("Variance: ",0.08276835709905783 - 0.181640852627094 )

Variance:  -0.09887249552803618


# #Building a KNeighbors Regression Model

In [66]:
from sklearn.neighbors import KNeighborsRegressor

In [67]:
KN_reg=KNeighborsRegressor(n_neighbors=5)  #creating an instance of the KNeighbors Regressor
KN_reg.fit(x_train,y_train)  #Traing our model using the training data set

KNeighborsRegressor()

In [68]:
KNreg_pred=KN_reg.predict(x_test) #testing how our model is doing using the unseen data

In [69]:
m.mean_squared_error(y_test,KNreg_pred) 

0.021139645948788043

In [70]:
m.r2_score(y_test,KNreg_pred)   # accuracy of model using our unseen data

0.5783116270498523

In [71]:
KNreg_score=m.r2_score(y_test,KNreg_pred)  #creating an instance  of the score of testing data set

In [72]:
m.mean_absolute_error(y_test,KNreg_pred)

0.0856645824842311

In [73]:
np.sqrt(m.mean_squared_error(y_test,KNreg_pred))  # Bias of our testing data set

0.14539479340329917

In [74]:
KNreg_pred_train=KN_reg.predict(x_train) # testing our model using the seen data

In [75]:
m.mean_squared_error(y_train,KNreg_pred_train)

0.015046965698313862

In [76]:
np.sqrt(m.mean_squared_error(y_train,KNreg_pred_train)) #Bias for our training data set

0.1226660739500285

In [77]:
m.r2_score(y_train,KNreg_pred_train)

0.6754477851954304

In [78]:
print('Variance: ',0.1226660739500285 - 0.14539479340329917)   #variance of the training and testing data

Variance:  -0.022728719453270665


# Creating a table of scores of the models

In [79]:
results={"Linear Regression": linear_score,"Ridge Regression":ridge_score,"Lasso Regression":lasso_score,"Decision Tree Regression":DecisionT_score,
             "KNeighbor Regression":KNreg_score}
final_results=pd.DataFrame(results, index=['Scores'])

In [80]:
final_results #printing a table that contains the summary of each model's accuracy

Linear Regression  Ridge Regression  Lasso Regression  \
Scores            0.64363          0.643631          0.590241   

        Decision Tree Regression  KNeighbor Regression  
Scores                  0.330268              0.578312

Comparing the accuracy of the Models

The above table has shown that the Ridge regression is more accurate going by the r2_score, hence it is the best model
amongst the five (5) regression algorithm i have considered. although, the LinearRegression is somewhat similar to the Ridge regression, it therefore means that using the linear regression can also serve

# PART 2

# GridSearchCV- RidgeRegression


In [81]:
from sklearn.model_selection import GridSearchCV

In [82]:
from sklearn.preprocessing import StandardScaler
#standard_x= StandardScaler()
#x=standard_x.fit_transform(x)

In [83]:
model=[]    #creating an empty list to contain our models
score=[]       #creating an empty list to handle our scores
parameter=[]    #creating an empty list to handle parameters

In [84]:
        #fine tuning hyperparameters

parameters = {'alpha': [0.0001,0.001,0.01,0.1]}   #choosing values for alpha

c1 = Ridge()   # Create an instance of the Ridge Regression model
grid = GridSearchCV(c1, parameters, cv=10)  #creating an instance of GridSearchCV with three parameters

grid.fit(x,y)    #Building the model 

print(grid.best_params_)     #printing the best fine tuned parameter

print(grid.best_estimator_)  #printing the best model using the best parameters

print(grid.best_score_)    #printing the best score for the model with the provided fine tuned parameters

model.append(c1)          #model[c1]

score.append(grid.best_score_)      #adding the best_score to the list 'score'

parameter.append(grid.best_estimator_)



{'alpha': 0.1}
Ridge(alpha=0.1)
0.5923258244155453


In [85]:
parameters = {'alpha':[0.0001,0.001,0.01,0.1]}   #fine tuning the parameters

c2 = Lasso()    #creating an instance of the lasso regression model

grid = GridSearchCV(c2, parameters, cv=10)

grid.fit(x,y)  #building the model 

print(grid.best_params_)    #printing the best parameter found during the GridSearchCV

print(grid.best_estimator_)    #printing the best estimator parameter

print(grid.best_score_)   #printing the best score for the model

model.append(c2)
score.append(grid.best_score_)
parameter.append(grid.best_estimator_)


{'alpha': 0.0001}
Lasso(alpha=0.0001)
0.5923234406187037


In [86]:
parameters = {'n_jobs':['-1','1'], 'fit_intercept':[True,False]} #defining the hyperparameters to be tuned

c3=LinearRegression()  #creating an instance of the LinearRegression model

grid = GridSearchCV(c3, parameters, cv=10) #creating an instance of the GridSearchCV using fine tuned parameters

grid.fit(x,y)   #building the model

#printing the best tuned paramters for the model
print(grid.best_params_)
print(grid.best_estimator_)
print(grid.best_score_)

#adding the models, best scores and best estimators into various lists
model.append(c3)
score.append(grid.best_score_)
parameter.append(grid.best_estimator_)



{'fit_intercept': True, 'n_jobs': '-1'}
LinearRegression(n_jobs='-1')
0.5923222918116509


In [87]:
parameters = {'n_neighbors': range(30),
             'metric':['manhattan','euclidean']} #creating a dictionary of parameters to be tuned

c4= KNeighborsRegressor()  #creating an instance of KNeighborsRegressor

grid = GridSearchCV(c4, parameters, cv=10) 

grid.fit(x,y)  #building the model with the hyperparameters

print(grid.best_params_)
print(grid.best_estimator_)
print(grid.best_score_)

model.append(c4)
score.append(grid.best_score_)
parameter.append(grid.best_estimator_)

C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\model_selection\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\model_selection\_validation.py", line 598, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\neighbors\_regression.py", line 190, in fit
    return self._fit(X, y)
  File "C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\neighbors\_base.py", line 514, in _fit
    raise ValueError(
ValueError: Expected n_neighbors > 0. Got 0

  warnings.warn("Estimator fit failed. The score on this train-test"
C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\model_selection\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these paramete

{'metric': 'euclidean', 'n_neighbors': 7}
KNeighborsRegressor(metric='euclidean', n_neighbors=7)
0.5415340685951453


C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\model_selection\_search.py:922: UserWarning: One or more of the test scores are non-finite: [       nan 0.2376627  0.4795917  0.50043584 0.52385047 0.53773393
 0.53818027 0.54113182 0.54088299 0.53878018 0.53540798 0.53181543
 0.53098026 0.52480755 0.52318729 0.51946631 0.51614705 0.51044293
 0.50211303 0.49894033 0.49130369 0.48600096 0.47835557 0.47129945
 0.46805621 0.46313074 0.45987757 0.45221372 0.4468179  0.44412926
        nan 0.2376627  0.4795917  0.50043584 0.5265686  0.5352243
 0.53862133 0.54153407 0.54049753 0.53952265 0.53532138 0.53114753
 0.52895201 0.52322667 0.52238177 0.51870376 0.5145989  0.50860955
 0.49949355 0.4961585  0.48898221 0.48525061 0.47627812 0.46941167
 0.46658264 0.46180233 0.45880605 0.45126304 0.44602562 0.44306937]
  warnings.warn(


In [88]:
parameters = {'criterion':['squared_error', 'friedman_mse','absolute error'],
             'splitter':['best','random']}   #buidling a dictionary of the parameters to be tuned

c5= DecisionTreeRegressor()   #creating an instance of DecisiontreeRegressor

grid = GridSearchCV(c5, parameters, cv=10)
grid.fit(x,y)       #building the model by using the hyperparameter

#printing the best hyperparameter, estimator and score fine tuned by GridSearchCV
print(grid.best_params_)
print(grid.best_estimator_)
print(grid.best_score_)

model.append(c5)
score.append(grid.best_score_)
parameter.append(grid.best_params_)

C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\model_selection\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\model_selection\_validation.py", line 598, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\tree\_classes.py", line 1252, in fit
    super().fit(
  File "C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\tree\_classes.py", line 351, in fit
    criterion = CRITERIA_REG[self.criterion](self.n_outputs_,
KeyError: 'squared_error'

  warnings.warn("Estimator fit failed. The score on this train-test"
C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\model_selection\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parame

{'criterion': 'friedman_mse', 'splitter': 'random'}
DecisionTreeRegressor(criterion='friedman_mse', splitter='random')
0.36418623790576654


C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\model_selection\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these parameters will be set to nan. Details: 
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\model_selection\_validation.py", line 598, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\tree\_classes.py", line 1252, in fit
    super().fit(
  File "C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\tree\_classes.py", line 351, in fit
    criterion = CRITERIA_REG[self.criterion](self.n_outputs_,
KeyError: 'absolute error'

  warnings.warn("Estimator fit failed. The score on this train-test"
C:\ProgramData\Anaconda3\conda\lib\site-packages\sklearn\model_selection\_validation.py:615: FitFailedWarning: Estimator fit failed. The score on this train-test partition for these param

In [89]:
final_table= pd.DataFrame({'MODEL NAME':model,'':'','BEST SCORE':score, 'BEST PARAMETERS':parameter})

In [90]:
print("SUMMARY OF MODELS".center(110),'\n')
print(final_table.to_string())

                                              SUMMARY OF MODELS                                                

                MODEL NAME    BEST SCORE                                         BEST PARAMETERS
0                  Ridge()      0.592326                                        Ridge(alpha=0.1)
1                  Lasso()      0.592323                                     Lasso(alpha=0.0001)
2       LinearRegression()      0.592322                           LinearRegression(n_jobs='-1')
3    KNeighborsRegressor()      0.541534  KNeighborsRegressor(metric='euclidean', n_neighbors=7)
4  DecisionTreeRegressor()      0.364186     {'criterion': 'friedman_mse', 'splitter': 'random'}


CONCLUSION

Part 1

The problem we are handling in this project is a regression problem, since we are trying to predict the charges based on certain variables, and charges is a continuous variable.

We have used five (5) different algorithms to develop a model to predict charges. From our model evaluation,  the Ridge regression has the best accuracy.the bias for our training data is 0.13879697981989667, while the bias for our testing data is 0.13366048111310602.

In my observation, the model is working almost same on both the testing and training data going by the rmse values 


Part 2

After building the model using GridSearchCV method and the five different model building algorithms and tuning the hyperparameters for the algorithm, a summary of my output is presented in the table above

From my observation, the Ridge Resgression with a score of  0.592326 is the best performing model.

In summary, my best performing model to predict charges is: cherges=(0.33061638)age + (0.01309255)children + (0.43375963)smoker  - (0.01546399)region + 0.07320254574145427